benchmarks/instruction_counts/definitions/standard.py

```diff
@@ -0,0 +1,72 @@
+"""Default set of benchmarks."""
+
+from core.api import GroupedModules, GroupedStmts
+from core.types import FlatIntermediateDefinition
+from core.utils import flatten, parse_stmts
+from definitions.setup import Setup
+
+
+BENCHMARKS: FlatIntermediateDefinition = flatten({
+    "Empty": {
+        "no allocation": GroupedStmts(
+            r"torch.empty(())",
+            r"torch::empty({0});",
```

Comment:
I'm wondering how this format will scale if you want to do really detailed, feature-by-feature instruction count benchmarking. For example, a while ago @swolchok was remarking how hitting the TensorOptions overload in empty was worse than the faithful overload, even when you were passing in a literal (e.g., `torch::empty({0}, at::kDouble)`) because C++ had to pack the literal into a TensorOptions, and then immediately unpack it to do the dispatch. The benchmark here isn't going to capture that; if you want to accurately capture the overheads of all the ways to call torch::empty, you're probably going to need a dozen or so variations (`torch::empty({0}, options)`, `torch::empty({0}, dtype)`, `torch::empty({0}, scalartype)`, etc). Looking at the current format, it seems that for every single line (or two, I guess, with the Python variant) of benchmark code, I need to write four more lines of boilerplate; that's more than doubling the size of the test in question. I think I, personally, would appreciate a much shorter format, esp, if I am going to have to do a lot of different variants.

Reply:
Based on the README, it also sounds like one of the things a lot of investment was made into was making it possible to run a benchmark under as many variations of the frontend as possible (so, C++, Python, TorchScript, etc.), but if I'm like painstakingly enumerating a lot of different torch.empty variations because I happen to know that these variations matter due to C++ implementation details, that doesn't necessarily mean that the TorchScript interpreter's performance faults cut along the same lines. So it might actually just be a waste of time to run each variation under TS, because actually you need a different set of variations to exercise different paths in the interpreter.

For microbenchmarks like this, I think it is legitimate to treating the system as a white box when deciding what to cover, because the point of instruction counts are to be actionable, and so when a benchmark fails we want this to correlate to some specific portion of code we ought to turn our attention to.

Reply:
Absolutely. Terse clear benchmark definitions are VERY important. I think generally the way as it currently stands there are two ways. So for instance in the example above:

`Option A: TimerArgs per line`

```
"Empty": {
    "no allocation": GroupedStmts(
        r"torch.empty(())",
        r"torch::empty({0});",
    ),
    "literal": TimerArgs("torch::empty({0}, at::kDouble);", language="cpp"),
    "TensorOptions": TimerArgs("torch::empty({0}, options);", "auto options = ...", language="cpp"),
    "ScalarType: TimerArgs("torch::empty({0}, scalartype);", "auto scalartype = ...", language="cpp"),
    ...
}
```

Pros:
* One line per def
* WYSIWYG
  
Cons:
* Fluff like `TimerArgs` and `language="cpp"` on each line.
* Nothing is aligned which makes it hard to read.

`Option B: Use comprehensions`

```
# This is a limitation of dicts. Without the `None` the later use of "Empty" would blow away this one and there's no way
# to detect and preserve it.
("Empty", None): {
    "no allocation": GroupedStmts(
        r"torch.empty(())",
        r"torch::empty({0});",
    ),
},
"Empty": {
    label: TimerArgs(f"torch::empty({{0}}, {args});", setup=setup, language="cpp")
    for label, args, setup in (
        ("literal", "at::kDouble", ""),
        ("TensorOptions", "options", "auto options = ..."),
        ("ScalarType", "scalartype", "auto scalartype = ..."),
        ...
    )
},
```

Pros:
* Scales very well. Signal to noise in the "definition" portion is nearly 100%
* Any sort of repeated code gets amortized out.
  
Cons:
* There is an O(5-10) line count overhead
* It will get very hard to read if one gets too clever. (True of all comprehensions, really.)

I'm imagining (A) would be for cases with 2-4 variants where you just don't want to pay the newline padding cost you describe, and (B) would be for where you have LOTS of cases but they're mechanical, and you can just read the comprehension args and get a very terse description of what is going on.

In the bigger set of benchmarks, I wound up using this sort of trick for indexing (https://github.com/pytorch/pytorch/pull/50608/files#diff-e4698f454a1a411efa34e153cedfe21dd06a129527fce585bfff813bdaa8455bR127-R152) and nn Modules (https://github.com/pytorch/pytorch/pull/50608/files#diff-e4698f454a1a411efa34e153cedfe21dd06a129527fce585bfff813bdaa8455bR224-R250) if you want to use that to gauge the "comprehension" approach to readable similar benchmarks.

But of course if you can come up with a more readable format I'm all ears.

Reply:
I saw the second comment after I wrote my response, but in summary: `TimerArgs` is the white box variant.

Reply:
I think the question is how to write things similar to what we have in tests:
```
for kwarg1, kwarg2, kwarg3 in product((True, False), (0,1,2), (torch.half, torch.float)):
     f(kwarg1=kwarg1, kwarg2=kwarg2, kwarg3=kwarg3)
```
I feel there must be a way to write it with a comprehension, the question is how readable and writeable it is

Reply:
@robieta just to clarify, in the A/B examples you gave above, the "no allocation" tests are just the existing benchmarks, correct? i.e. if I'm just doing something local I can drop them. and it's just a choice between e.g.
```
{
    "literal": TimerArgs("torch::empty({0}, at::kDouble);", language="cpp"),
    "TensorOptions": TimerArgs("torch::empty({0}, options);", "auto options = ...", language="cpp"),
    "ScalarType: TimerArgs("torch::empty({0}, scalartype);", "auto scalartype = ...", language="cpp"),
    ...
}
```
and
```
{
    label: TimerArgs(f"torch::empty({{0}}, {args});", setup=setup, language="cpp")
    for label, args, setup in (
        ("literal", "at::kDouble", ""),
        ("TensorOptions", "options", "auto options = ..."),
        ("ScalarType", "scalartype", "auto scalartype = ..."),
        ...
    )
}
```
If so, then how far from reality is this transcription of Natalia's example? Judging by the cardinality of `label` in your example it doesn't seem like comprehensions have the same combinatorial behavior as `product` - if that's accurate, would it be feasible to add? (We'd need some way to generate the necessary `label` entries ofc but I'm not sure what else would be needed)
```
TimerArgs(f"f(kwarg1={kwarg1}, kwarg2={kwarg2}, kwarg3={kwarg3});", 
    for kwarg1, kwarg2, kwarg3 in ((true, false), (0, 1, 2), (torch.half, torch.float))
)
```

Reply:
The emphasis on list comprehensions still feels a little off to me.

Let me describe my thought process for how I would setup microbenchmarks for, say torch::empty.

For torch::empty, I'm primarily interested in dispatcher overhead, and exercising the different overloads. So when writing this benchmark I take a look at the empty overloads (ignoring named tensors for the moment):

```
TORCH_API Tensor empty(IntArrayRef size, TensorOptions options={}, c10::optional<MemoryFormat> memory_format=c10::nullopt);
TORCH_API Tensor empty(IntArrayRef size, c10::optional<ScalarType> dtype, c10::optional<Layout> layout, c10::optional<Device> device, c10::optional<bool> pin_memory, c10::optional<MemoryFormat> memory_format);
```

I want to exercise each of these overloads. So I sit down and start writing down calls which express the different overloads:

```
int main() {

  // Setup
  auto options_empty = c10::TensorOptions();
  auto options_full = c10::TensorOptions().dtype(kFloat).device(kCPU);
  auto optional_float = c10::make_optional(kFloat);

  // TensorOptions overload
  at::empty({0}, options_empty);
  at::empty({0}, options_full);
  at::empty({0}, at::kFloat); // implicit conversion

  // Faithful overload
  at::empty({0}, nullopt, nullopt, nullopt, nullopt, nullopt);
  at::empty({0}, kFloat, nullopt, nullopt, nullopt, nullopt);
  at::empty({0}, at::optional_float, nullopt, nullopt, nullopt, nullopt);

  // Baseline comparison
  at::empty({0});
  at::cpu::empty({0});

  return 0;
}
```

I like that it's a little C++ program, since that means I can paste it into Godbolt and look at the assembly easily. Because I'm whiteboxing, I'm not interested in the full combinatorial explosion of possibilities (e.g., I don't test everything on both at::cpu and at::), because I happen to know that the relevant TensorOptions unpacking code is shared between the two implementations, so one comparison is good enough. Because I'm specifically interested in the overload handling, I purposely don't exercise differing sizes; that's benchmarking something else, and maybe I have a separate test program for that.

Now, I need to translate this into benchmarks inside the framework. Honestly, this program (maybe with some small modifications, e.g., no int main wrapper) ought to be enough:

1. I don't want to have to name my individual benchmarks. Each line of code is super self explanatory, and should be translated into the name itself. Maybe I give an explanatory comment for one or subtle cases.
2. I don't want to have to manage propagating the setup lines to the appropriate benchmark that uses it. I define one setup block and use it for all the benchmarks. Maybe if the setup is time consuming and I don't need it for some other benchmark, I DCE the useless setup lines.
3. I don't want to have to setup a list comprehension. One line, one benchmark.

Reply:
I think what it is that I really hate about list comprehensions is, from the reader's perspective, it obscures what the *actual* C++ code that is being run under the benchmark is. I don't mind abstracting away setup code, but for the actual thing I'm measuring, I really *do* want to see the code in question, I don't want it DRYed away

Reply:
@ezyang @bhosmer @ngimel I've added a `GroupedVariants` definition scheme, which basically implements the "benchmark-per-line" approach discussed earlier. It also seems to compose reasonably well with `parse_stmts`; for instance indexing benchmarks (in the other stack) weren't particularly ergonomic with comprehensions, whereas now it's much clearer what's going on.